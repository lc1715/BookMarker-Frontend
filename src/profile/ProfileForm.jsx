import { useState, useContext } from 'react';
import ShowAlert from '../common/ShowAlert';
import BookMarkerApi from '../api/api';
import UserContext from '../auth/UserContext';


/** Edit profile form.
 * 
 * Route: /profile
 * Displays profile form
 * Submitting the form calls the BookMarkerApi to update user profile
 * If successful, triggers current user to be reloaded throughout the site and
 * shows successful alert message
 */
function ProfileForm() {
    const { currentUser, setCurrentUser } = useContext(UserContext);

    const [formData, setFormData] = useState({
        username: currentUser.username,
        email: currentUser.email
    });

    const [formErrors, setFormErrors] = useState([]);
    const [isSaved, setIsSaved] = useState(false);


    /** on form submit:
     * - attempt to save new user info to backend & report any errors
     * - if successful
     *   - clear previous error messages
     *   - show successful alert message
     *   - set new current user info throughout the site
     */
    async function handleSubmit(evt) {
        evt.preventDefault();
        let username = formData.username;
        let newEmail = formData.email;
        let updatedUser;

        try {
            updatedUser = await BookMarkerApi.updateUserProfile(username, { email: newEmail });
        } catch (err) {
            setFormErrors(err);
            return;
        }

        setFormErrors([]);
        setIsSaved(true);
        // trigger reloading of current user information throughout the site
        setCurrentUser(updatedUser);

        setTimeout(() => {
            setIsSaved(false)
        }, 3000)
    };

    /** Handle form info changing*/
    function handleChange(evt) {
        const { name, value } = evt.target;
        setFormData(data => ({ ...data, [name]: value }));
        setFormErrors([]);
    }

    return (
        <form onSubmit={handleSubmit}>
            <label htmlFor='username'>Username</label>
            <input
                id='username'
                name='username'
                value={formData.username}
                disabled={true}
            />

            <label htmlFor='email'>Email</label>
            <input
                id='email'
                name='email'
                type='email'
                value={formData.email}
                onChange={handleChange}
            />

            {formErrors.length ? <ShowAlert type='danger' messages={formErrors} /> : null}

            {isSaved ? <ShowAlert type='success' messages={['Updated successfully']} /> : null}

            <button>Submit</button>
        </form>

    )
};

export default ProfileForm;