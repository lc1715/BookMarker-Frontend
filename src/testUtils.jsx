import UserContext from "./auth/UserContext";

const demoUser = {
    id: 1,
    username: 'testUser',
    email: 'test@email.com',
    volume_ids: ['e3', 't5'],
}

const UserProvider =
    ({ children, currentUser = demoUser, hasSavedBook = () => (true) }) => (
        <UserContext.Provider value={{ currentUser, hasSavedBook }}>
            {children}
        </UserContext.Provider>
    );

export { UserProvider };
