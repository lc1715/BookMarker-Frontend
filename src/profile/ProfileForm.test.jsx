import { it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from "../testUtils";
import ProfileForm from './ProfileForm';

it('renders the Profile Form component', () => {
    render(
        <MemoryRouter>
            <UserProvider>
                <ProfileForm />
            </UserProvider>
        </MemoryRouter>
    )
})

it('matches with snapshot', () => {
    const { asFragment } = render(
        <MemoryRouter>
            <UserProvider>
                <ProfileForm />
            </UserProvider>
        </MemoryRouter>
    )

    expect(asFragment()).toMatchSnapshot();
})

it('should correctly render the text User Profile', () => {
    render(
        <MemoryRouter>
            <UserProvider>
                <ProfileForm />
            </UserProvider>
        </MemoryRouter>
    )

    expect(screen.getByText('User Profile')).toBeInTheDocument()
})