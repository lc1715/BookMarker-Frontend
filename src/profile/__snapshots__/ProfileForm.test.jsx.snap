// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches with snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm css-qd6y3d-MuiContainer-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation4 css-1i3pr2w-MuiPaper-root"
        style="--Paper-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="MuiBox-root css-1ykdma4"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-19hrv1-MuiTypography-root"
          >
            User Profile
          </p>
        </div>
        <div
          class="MuiBox-root css-1rp7mde"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-w7xktn-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-disabled MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-113d811-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r2:"
              id=":r2:-label"
            >
              Username
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                disabled=""
                id=":r2:"
                name="username"
                type="text"
                value="testUser"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-w1u3ce"
                >
                  <span>
                    Username
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1rp7mde"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-w7xktn-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-113d811-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r3:"
              id=":r3:-label"
            >
              Email
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r3:"
                name="email"
                type="email"
                value="test@email.com"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-w1u3ce"
                >
                  <span>
                    Email
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1ykdma4"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1e88gys-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Submit
          </button>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
